Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/transport.go

```diff
@@ -17,16 +17,174 @@ limitations under the License.
+}
+
+func NewRequestInfo(req *http.Request, readBody bool) *RequestInfo {
+	reqInfo := &RequestInfo{}
+	reqInfo.RequestURL = req.URL.String()
+	reqInfo.RequestVerb = req.Method
+	reqInfo.RequestHeaders = req.Header
+	if readBody && req.Body != nil {
+		if body, err := ioutil.ReadAll(req.Body); err == nil {
+			reqInfo.RequestBody = body
+			req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		}
+	}
+
+	return reqInfo
+}
+
+func (r *RequestInfo) Complete(response *http.Response, err error, readBody bool) {
+	if err != nil {
+		r.ResponseErr = err
+		return
+	}
+	r.ResponseStatus = response.Status
+	r.ResponseHeaders = response.Header
+	if readBody && response.Body != nil {
+		if body, err := ioutil.ReadAll(response.Body); err == nil {
+			r.ResponseBody = body
+			response.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		}
+	}
+}
+
+func (r RequestInfo) ToCurl() string {
+	headers := ""
+	for key, values := range map[string][]string(r.RequestHeaders) {
+		for _, value := range values {
+			headers += fmt.Sprintf(` -H %q`, fmt.Sprintf("%s: %s", key, value))
+		}
+	}
+
+	body := ""
+	if len(r.RequestBody) > 0 {
+		body = fmt.Sprintf("-d %q", string(body))
+	}
+
+	return fmt.Sprintf("curl -k -v -X%s %s %s %s", r.RequestVerb, headers, body, r.RequestURL)
+}
+
+// TrackingRoundTripper keeps track of all the requests made.  You should use this with caution, because it grow with every request.
+type TrackingRoundTripper struct {
+	delegatedRoundTripper http.RoundTripper
+
+	RequestInfos []RequestInfo
+}
+
+func NewTrackingRoundTripper(rt http.RoundTripper) *TrackingRoundTripper {
+	return &TrackingRoundTripper{rt, []RequestInfo{}}
+}
+
+func (rt *TrackingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	reqInfo := NewRequestInfo(req, bool(glog.V(7)))
+
+	startTime := time.Now()
+	response, err := rt.delegatedRoundTripper.RoundTrip(req)
+	reqInfo.Duration = time.Since(startTime)
+
+	reqInfo.Complete(response, err, bool(glog.V(7)))
+	rt.RequestInfos = append(rt.RequestInfos, *reqInfo)
+
+	return response, err
+}
+
+// DebuggingRoundTripper will display information about the requests passing through it based on what is configured
+type DebuggingRoundTripper struct {
+	delegatedRoundTripper http.RoundTripper
+
+	Levels util.StringSet
+}
+
+const (
+	JustURL         string = "url"
+	URLTiming       string = "urltiming"
+	CurlCommand     string = "curlcommand"
+	RequestBody     string = "requestbody"
+	RequestHeaders  string = "requestheaders"
+	ResponseStatus  string = "responsestatus"
+	ResponseBody    string = "responsebody"
+	ResponseHeaders string = "responseheaders"
 )
 
+func NewDebuggingRoundTripper(rt http.RoundTripper, levels ...string) *DebuggingRoundTripper {
+	return &DebuggingRoundTripper{rt, util.NewStringSet(levels...)}
+}
+
+func (rt *DebuggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	reqInfo := NewRequestInfo(req, rt.Levels.Has(RequestBody))
+
+	if rt.Levels.Has(JustURL) {
```

Comment:
> do you know how the lookup compares to a bitmask? wondering about 8 map lookups per http request on something this low-level

shrugs.  It's an optional level in the transport chain.  If you care about cost, don't add it.